server/channels/app/oauthproviders/gitlab/gitlab.go

```diff
@@ -20,24 +20,31 @@ type GitLabProvider struct {
 }
 
 type GitLabUser struct {
-	Id       int64  `json:"id"`
-	Username string `json:"username"`
-	Login    string `json:"login"`
-	Email    string `json:"email"`
-	Name     string `json:"name"`
+	Id                int64  `json:"id"`
+	Username          string `json:"username"`
+	Login             string `json:"login"`
+	Email             string `json:"email"`
+	Name              string `json:"name"`
+	PreferredUsername string `json:"preferred_username"`
 }
 
 func init() {
 	provider := &GitLabProvider{}
 	einterfaces.RegisterOAuthProvider(model.UserAuthServiceGitlab, provider)
 }
 
-func userFromGitLabUser(logger mlog.LoggerIFace, glu *GitLabUser) *model.User {
+func userFromGitLabUser(logger mlog.LoggerIFace, glu *GitLabUser, settings *model.SSOSettings) *model.User {
 	user := &model.User{}
 	username := glu.Username
 	if username == "" {
 		username = glu.Login
 	}
+
+	if settings != nil && model.SafeDereference(settings.UsePreferredUsername) && glu.PreferredUsername != "" {
+		// to maintain consistency with other providers, we split the preferred username by @ and take the first part
```

Comment:
Oh, I this comment isn't on the enterprise side of things, so I didn't realize this was the intent -- makes sense to split and clean always then in that PR.

But wondering if structure could be simplified as also proposed there?

Reply:
Good idea, done!